exonum-java-binding/core/src/test/java/com/exonum/binding/core/storage/database/PrefixedIntegrationTest.java

```diff
@@ -0,0 +1,154 @@
+          .getProofEntry(IndexAddress.valueOf(entryName), string());
+
+      // Check it is the same instance (which is required for correct operation with Forks):
+      assertThat(e2).isSameAs(e1);
+      // fixme: There is an issue with such caches: #getIndexAddress _may_ return full address
+      //   (if (a) the prefixed is created with a Fork; (b) the index is cached there),
+      //   or the same address that the user passed (if the index is not cached, or the prefixed
+      //   was created from handle
+    }
+
+    @Test
+    void fromForkAccessPrefixedFirst() {
+      // Create the base access — a Fork
+      Fork fork = db.createFork(cleaner);
+
+      // Create a Prefixed Access to that namespace
+      Prefixed prefixed = Prefixed.fromAccess(namespace, fork);
+      // Access the index from the Prefixed
+      ProofEntryIndexProxy<String> e1 = prefixed
+          .getProofEntry(IndexAddress.valueOf(entryName), string());
+
+      // Try to access the same index from the Fork
+      String fullName = namespace + "." + entryName;
+      ProofEntryIndexProxy<String> e2 = fork
+          .getProofEntry(IndexAddress.valueOf(fullName), string());
+
+      // Check it is the same instance (which is required for correct operation with Forks):
+      assertThat(e2).isSameAs(e1);
+    }
+
+    @Test
+    void fromReadonlyAccessSnapshotBaseFirst() {
+      // Create the base access — a Snapshot
+      Snapshot base = db.createSnapshot(cleaner);
+
+      // Access the index from the base
+      String fullName = namespace + "." + entryName;
+      ProofEntryIndexProxy<String> e1 = base
+          .getProofEntry(IndexAddress.valueOf(fullName), string());
+
+      // Create a Prefixed Access to that namespace
+      Prefixed prefixed = Prefixed.fromAccess(namespace, base);
+      // Try to access the same index from the Prefixed
+      ProofEntryIndexProxy<String> e2 = prefixed
+          .getProofEntry(IndexAddress.valueOf(entryName), string());
+
+      // Check the entry has the same _state_: readonly accesses do not require caching
+      assertThat(e1.get()).isEqualTo(e2.get());
+    }
+
+    @Test
+    void baseCleanerInvalidatesIndexesFromPrefixed() throws CloseFailuresException {
+      // Create the base access — a Snapshot
+      Snapshot base = db.createSnapshot(cleaner);
+
+      // Create a Prefixed Access from the base
+      Prefixed prefixed = Prefixed.fromAccess(namespace, base);
+      // Create an index from the Prefixed
+      ProofEntryIndexProxy<String> index = prefixed
+          .getProofEntry(IndexAddress.valueOf(entryName), string());
+
+      // Close the cleaner
+      cleaner.close();
+
+      // Check both the access and the index are inaccessible
+      assertThrows(IllegalStateException.class, prefixed::getAccessNativeHandle);
+      assertThrows(IllegalStateException.class, index::get);
+    }
+
+    @AfterEach
+    void dropDatabase() throws CloseFailuresException {
+      try {
+        cleaner.close();
```

Comment:
NPE if an exception occurs in creating db 🙃

Reply:
d5daac96f772cb8ac5c128d91cf757f31a159d58